### PR TITLE
Added pyversion option for doctest (issue 3303)

### PR DESCRIPTION
**Pull request by marco-buttu**
This pull request is related to the   #3303 . Now it is possible to specify the Python version required to execute the doctest. For instance:

```
.. doctest::
   :pyversion: >= 3.3
```

The operator could be: <, <=, ==, >=, >. I also added two tests.

### CODE REVIEW

File: sphinx/ext/doctest.py [hunk 1]
```diff
@@ -23,6 +24,8 @@
 from docutils import nodes
 from docutils.parsers.rst import Directive, directives
 
+from distutils.version import StrictVersion as V
```
Comment (tk0miya) [inline sphinx/ext/doctest.py]: Why abbrviate? `V` is not clear name to me.

Anyway, it should be imported after importing six.
We usually import libraries in order of python standards, 3rd parties, docutils and sphinx's.

File: sphinx/ext/doctest.py [hunk 2]
```diff
@@ -103,10 +106,40 @@ def run(self):
...
                         doctest.OPTIONFLAGS_BY_NAME):  # type: ignore
-                    # XXX warn?
+                    self.state.document.reporter.warning(
+                            "missing '+' or '-' in '%s' option." % option, 
+                            line=self.lineno)
                     continue
                 flag = doctest.OPTIONFLAGS_BY_NAME[option[1:]]  # type: ignore
                 node['options'][flag] = (option[0] == '+')
+        if self.name == 'doctest' and 'pyversion' in self.options:
+            try:
```
Comment (tk0miya) [inline sphinx/ext/doctest.py:144]: Could you separate this into a single function and add some test?
It seems the evaluation code are not tested well.

**Review by tk0miya — COMMENTED** [review]
LGTM with nits

**Comment by tk0miya** [comment]
Ah, I'd overlooked about travis.
It seems CI process was failed.
Please check the logs of travis ci and fix them.

**Comment by marco-buttu** [comment]
Thanks for your time @tk0miya . I am not completely satisfied about the test, because I do not know how to propery create a TestDirective object, so I gave it dummy arguments...

File: sphinx/ext/doctest.py [hunk 3]
```diff
@@ -107,41 +106,52 @@ def run(self):
...
+        This function is meant to be used to evaluate the doctest :pyversion:
+        option. For instance, if the doctest directive provides the option
+        :pyversion: >= 3.3, then we have to check if the running Python version
+        is greather or equal to 3.3. In that case, operand will be the string
+        '>=' and version the string '3.3'. proper_pyversion() will return True
+        if the running Python version is >= 3.3, False otherwise.
+        """
+        operands = ('<=', '<', '==', '>=', '>')
+        if operand not in operands:
+            self.document.reporter.warning(
```
Comment (tk0miya) [inline sphinx/ext/doctest.py]: Is this typo for self. **state** .reporter?
Did you check this warning works surely?

**Review by tk0miya — COMMENTED** [review]
If you are not satisfied the tests, how about changing the interface like following?
```
def compare_version(ver1, ver2, operand):
    if operand not in operands:
        raise ValueError  # raise ValueError if unknown operand

    ...
    return result  # return boolean value as a result of comparison


class TestDirective(Directive):
    def run(self):
        ...
        if self.name == 'doctest' and 'pyversion' in self.options:
            try:
                if compare_version(platform.python_version(), version, operand):
                    ...
            except ValueError:
                # handle error for unknown operand and other reasons.
                self.state.document.reporter.warning(...)
```

File: sphinx/ext/doctest.py (hunk 3 above)
Comment (marco-buttu) [reply sphinx/ext/doctest.py]: Your interface is just perfect :-)  So, I will change my interface following your suggestion, fix the typo and check (with pytest) everything works properly. Thank you very much!

File: doc/ext/doctest.rst [hunk 4]
```diff
@@ -73,6 +73,19 @@ a comma-separated list of group names.
      explicit flags per example, with doctest comments, but they will show up in
      other builders too.)
 
+   * ``pyversion``, a string option, can be used to specify the required Python
+     version for the example to be tested. For instance, in the following case
+     the example will be tested only for Python versions greather than 3.3::
+
+         .. doctest::
+            :pyversion: > 3.3
+
```
Comment (tk0miya) [inline doc/ext/doctest.rst]: style-check script says
```
doc/ext/doctest.rst:82: trailing whitespace
```

Please check Travis CI's status after fixed up.

**Review by tk0miya — APPROVED** [review]
LGTM with nits

File: doc/ext/doctest.rst (hunk 4 above)
Comment (marco-buttu) [reply doc/ext/doctest.rst]: I am sorry @tk0miya , I checked Travis but I was not able to figure out where the problem was, and I thought to fix it on this morning. I have not enough confidence with all the tools yet :-/

**Comment by marco-buttu** [comment]
I double checked everything, the warning messages too, and it works properly. Thank you very very much @tk0miya

**Comment by tk0miya** [comment]
LGTM!
I will merge this if no objection in a few days.

File: CHANGES [hunk 5]
```diff
@@ -73,6 +73,8 @@ Deprecated
 Features added
 --------------
 
+* #3303: Added ``:pyversion:`` option to the doctest directive
```
Comment (tk0miya) [inline CHANGES]: Sorry, this feature will be released as 1.6.
1.5.x series is under maintenance period. so only bug fixes are allowed.

File: doc/ext/doctest.rst [hunk 6]
```diff
@@ -73,6 +73,19 @@ a comma-separated list of group names.
...
+     The supported operands are ``<``, ``<=``, ``==``, ``>=``, ``>``, and
+     comparison is performed by `distutils.version.LooseVersion
+     <https://www.python.org/dev/peps/pep-0386/#distutils>`__.
+
+     .. versionadded:: 1.5
```
Comment (tk0miya) [inline doc/ext/doctest.rst]: Please rewrite this to 1.6.

File: sphinx/ext/doctest.py [hunk 7]
```diff
@@ -101,12 +126,32 @@ def run(self):
             # parse doctest-like output comparison flags
             option_strings = self.options['options'].replace(',', ' ').split()
             for option in option_strings:
-                if (option[0] not in '+-' or option[1:] not in
-                        doctest.OPTIONFLAGS_BY_NAME):  # type: ignore
-                    # XXX warn?
+                on_or_off, option_name = option[0], option[1:]
+                if on_or_off not in '+-':  # type: ignore
+                    self.state.document.reporter.warning(
+                        "missing '+' or '-' in '%s' option." % option,
```
Comment (tk0miya) [inline sphinx/ext/doctest.py]: Could you wrap with `_(...)` function?
Ex:
```
from sphinx.locale import _
self.state.document.reporter.warning(_("missing '+' or '-' in '%s' option.") % option)
```

Then it makes the message translatable.

File: sphinx/ext/doctest.py [hunk 8]
```diff
@@ -101,12 +126,32 @@ def run(self):
             # parse doctest-like output comparison flags
             option_strings = self.options['options'].replace(',', ' ').split()
             for option in option_strings:
-                if (option[0] not in '+-' or option[1:] not in
-                        doctest.OPTIONFLAGS_BY_NAME):  # type: ignore
-                    # XXX warn?
+                on_or_off, option_name = option[0], option[1:]
```
Comment (tk0miya) [inline sphinx/ext/doctest.py]: I feel `on_or_off` is a bit ugly name. How about `prefix`, `sign` or others?

**Review by tk0miya — CHANGES_REQUESTED** [review]
I reviewed again for latest update.
Could you check them please?

File: sphinx/ext/doctest.py (hunk 8 above)
Comment (marco-buttu) [reply sphinx/ext/doctest.py]: ``on_or_off`` is the name given to ``+/-`` in the doctest documentation, but I agree with you that is not meaningful. I wil use ``prefix``, thanks for the suggestion :-)

File: sphinx/ext/doctest.py (hunk 7 above)
Comment (marco-buttu) [reply sphinx/ext/doctest.py]: :+1:

File: doc/ext/doctest.rst (hunk 6 above)
Comment (marco-buttu) [reply doc/ext/doctest.rst]: :+1:

**Comment by marco-buttu** [comment]
Done :) Thanks @tk0miya

File: sphinx/ext/doctest.py (hunk 8 above)
Comment (tk0miya) [reply sphinx/ext/doctest.py]: I don't know that. But it looks better to me :-)

**Review by tk0miya — APPROVED** [review]

**Comment by tk0miya** [comment]
Merged!
Thank you for contribution :-)

**Pull request merged**